services/workflows-service/src/collection-flow/collection-flow.service.ts

```diff
@@ -229,6 +238,45 @@ export class CollectionFlowService {
     return flowData;
   }
 
+  private async __persistFileUrlsToDocuments(
+    documents: TDocument[] = [],
+    projectIds: TProjectIds,
+  ): Promise<TDocument[]> {
+    const fileEntities = await Promise.all(
+      documents.reduce((filesList, document) => {
+        document.pages.forEach((page: { ballerineFileId: string }) => {
+          filesList.push(this.storageService.getFileById({ id: page.ballerineFileId }, projectIds));
+        });
+
+        return filesList;
+      }, [] as Promise<File | null>[]),
+    );
+
+    const filesById = keyBy(fileEntities, 'id');
+
+    const updatedDocuments = documents.map(document => {
+      return {
+        ...document,
+        pages: document.pages.map(
+          (page: { ballerineFileId: string; uri: string; provider: string; type: string }) => {
+            const file: File | null = filesById[page.ballerineFileId] ?? null;
```

Comment:
As long as we return, throw an error, or use `!` over `ballerineFileId` being `undefined` types are being inferred here as `File | null` as expected.

```suggestion
            const file = filesById[page.ballerineFileId!] ?? null;
```